### PR TITLE
Fix some bugs in dyno scope resolver by thunking builders

### PR DESCRIPTION
**Pull request by DanilaFe**
There's a buggy interaction in the the `convert-uast` pass between builders that clone their nodes (e.g. those for `coforall` and `local`) and fixups performed via `TemporaryConversionSymbol`. The temporary conversion symbol is eventually replaced with a "proper" symbol using an `ID`-based lookup. However, cloning performed by the various builders means that the same `ID` in the uAST can actually correspond to multiple `Symbol` nodes in the production AST.  All `TemporaryConversionSymbol`s for a particular `ID` end up pointing to a particular `Symbol`, rather than their respective copies. This manifests as use-before-definition errors, but in reality it creates an AST in which one branch of an if-expression contains a variable that refers to a definition in another branch.

The solution to this problem is to avoid cloning until `TemporaryConversionSymbol`s have been replaced with proper symbols (i.e., until after fixups have been applied). We could do this by modifying the various builders to shuffle around the copying to work properly with `convert-uast` and its fixup machinery. However, in this PR, I propose an alternative approach: using [thunks](https://en.wikipedia.org/wiki/Thunk).

A thunk is a basically a deferred unit of computation. In the case of this PR, a thunk is a call to a builder, as well as the arguments to that builder, that _haven't yet happened, but will_. For instance, the following code changes occur.
```C++
    // Old
    return buildCoforallLoopStmt(indices, iterator, byref_vars, body, zippered);
    // New
    return buildThunk(buildCoforallLoopStmt, indices, iterator, byref_vars, body, zippered);
```
The difference between the old and new version is that the new version _does not yet invoke the builder_. Its arguments are saved, and it creates an AST node (a `Thunk*`); in other words, this new `Thunk` is just another node that can be traversed, visited, and have its children be fixed up. Eventually, thunks can be made to evaluate the function they stand for (`buildCoforallLoopStmt` in this case), by calling `force`. This, however, can be made to happen _after_ fixups are applied, and thus, the builder can do all the copying and mangling it wants safely.

In this PR, I use the thunk to fix bugs caused by copying in `local` and `coforall` statements. Any builder than is copied can be wrapped in a `thunk` to mitigate the effects of its copying.

Reviewed by @mppf - thanks!

## Testing
- [x] Paratest
- [x] Paratest with --dyno and updates contained in #21596 (down to 11 failures, 7 of which are `parseTwice`)

### CODE REVIEW

File: compiler/AST/Thunk.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  if (visitor->enterThunk(this)) {
+    for_alist(child, children) {
+      child->accept(visitor);
+    }
+    visitor->exitThunk(this);
+  }
+}
+
+Thunk* Thunk::copyInner(SymbolMap* map) {
+  INT_FATAL("thunks should not be copied");
```
Comment (mppf) [inline compiler/AST/Thunk.cpp:44]: Hmm, I wonder if you could end up with this, with nested coforalls or something?

File: compiler/include/Thunk.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  comiler's AST. The production comiler AST node builders do some mangling of
+  their arguments, in some cases cloning them. However, cloning breaks the
+  convert-uast pass' "fixups". A "fixup" might be used to point a SymExpr
+  to a Symbol that's only syntactically defined later, and thus not available
+  during the initial conversion pass.
+
+  To work around this, the thunk represents a not-yet-built AST node that
+  nevertheless already has all the arguments required for it to build. Its
+  children are accessible using normal iteration, but only the `force` method
+  will actually cause the "proper" builder to be invoked. This way, the
```
Comment (mppf) [inline compiler/include/Thunk.h:37]: let's leave out the quotes around proper

File: compiler/include/Thunk.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  mangling builders can be deferred until after fixups are complete.
+
+  This Thunk is a base class, extended by various templated versions to make
+  the API more argonomic.
```
Comment (mppf) [inline compiler/include/Thunk.h:41]: ergonomic

File: compiler/resolution/resolveFunction.cpp [hunk 4]
```diff
@@ -45,6 +45,7 @@
 #include "stmt.h"
 #include "stringutil.h"
 #include "symbol.h"
+#include "Thunk.h"
```
Comment (mppf) [inline compiler/resolution/resolveFunction.cpp:48]: is this needed here?

File: compiler/include/baseAST.h [hunk 5]
```diff
@@ -146,6 +148,8 @@ typedef std::vector<NameAndSymbol> SymbolNameVec;
 // enumerated type of all AST node types
 //
 enum AstTag {
+  E_Thunk,
```
Comment (mppf) [inline compiler/include/baseAST.h:151]: I would expect it to be added next to TemporaryConversionSymbol in this enum rather than as the 1st thing

File: compiler/include/baseAST.h [hunk 6]
```diff
@@ -188,7 +192,7 @@ enum AstTag {
 };
 
 static inline bool isExpr(AstTag tag)
-{ return tag >= E_SymExpr        && tag <= E_ExternBlockStmt; }
+{ return tag >= E_Thunk          && tag <= E_ExternBlockStmt; }
```
Comment (mppf) [inline compiler/include/baseAST.h:195]: If you move E_Thunk next to E_TemporaryConversionSymbol as I recommended above, you will need to undo this change.

**Review by mppf — APPROVED** [review]
This is looking good, along with changes we talked about elsewhere -- in particularly renaming it to TemporaryConversionThunk or something like that.

File: compiler/include/Thunk.h (hunk 3 above)
Comment (DanilaFe) [reply compiler/include/Thunk.h:41]: 🤦

File: compiler/include/baseAST.h (hunk 5 above)
Comment (DanilaFe) [reply compiler/include/baseAST.h:151]: Well, the thunk is an expr and the TCS is a symbol, so I was trying to keep them separated. I can definitely keep the thunk at the end of the expr / stmt list, which would put it very close to (barring a space) the TCS.

Comment (mppf) [reply compiler/include/baseAST.h:151]: Sounds good thanks

File: compiler/resolution/resolveFunction.cpp (hunk 4 above)
Comment (DanilaFe) [reply compiler/resolution/resolveFunction.cpp:48]: yes, for the various macros that do something over each AST node type (e.g., "for each child").

File: compiler/AST/Thunk.cpp (hunk 1 above)
Comment (DanilaFe) [reply compiler/AST/Thunk.cpp:44]: My thinking was that copying is deferred until thunks are forced, and that thunks would be forced in a "deeper-first" order (because we convert the AST "bottom up"). This would mean that by the time a thunk is forced, it should contain no inner thunks. I'll check that nested coforall cases don't break, though.

Comment (DanilaFe) [reply compiler/AST/Thunk.cpp:44]: The following works fine under `--dyno --verify`:
```Chapel
class C {}
coforall 1..10 {
    coforall 1..10 {
        if var x = new C() {
            x;
        }
    }
}
```
